pdfrw/objects/pdfstring.py

```diff
@@ -1,74 +1,547 @@
+            if bytes_encoding == 'auto' and len(splitlist) // 2 >= len(raw):
+                force_hex = True
+
+        if force_hex:
+            # The spec does not mandate uppercase,
+            # but it seems to be the convention.
+            fmt = '<%s>'
+            result = binascii.hexlify(raw).upper()
+        else:
+            fmt = '(%s)'
+            splitlist[1::2] = [(b'\\' + x) for x in splitlist[1::2]]
+            result = b''.join(splitlist)
+
+        return cls(fmt % convert_load(result))
+
+    @classmethod
+    def from_unicode(cls, source, text_encoding='auto',
+                     bytes_encoding='auto'):
+        """ The from_unicode() constructor is called to encode a source
+            string into a PdfString that is suitable for inclusion in a PDF.
+
+            NOTE:  There is no magic in the encoding process.  A user
+            can certainly do his own encoding, and simply initialize a
+            PdfString() instance with his encoded string.  That may be
+            useful, for example, to add line breaks to make it easier
+            to load PDFs into editors, or to not bother to escape balanced
+            parentheses, or to escape additional characters to make a PDF
+            more readable in a file editor.  Those are features not
+            supported by this method.
+
+            from_unicode() can use a heuristic to figure out the best
+            encoding for the string, or the user can control the process
+            by changing the text_encoding parameter to 'pdfdocencoding'
+            or 'utf16', and/or by changing the bytes_encoding parameter
+            to 'literal' or 'hex' to force particular conversion methods.
+
+            The function will raise an exception if it cannot perform
+            the conversion as requested by the user.
+        """
+
+        # Give preference to pdfdocencoding, since it only
+        # requires one raw byte per character, rather than two.
+        if text_encoding != 'utf16':
+            force_pdfdoc = text_encoding == 'pdfdocencoding'
+            if text_encoding != 'auto' and not force_pdfdoc:
+                raise ValueError('Invalid text_encoding value: %s'
+                                 % text_encoding)
+
+            if not source.startswith(cls.uni_bom):
+                try:
+                    raw = source.encode('pdfdocencoding')
+                except UnicodeError:
+                    if force_pdfdoc:
+                        raise
+                else:
+                    return cls.from_bytes(raw, bytes_encoding)
+
+        # If the user is not forcing literal strings,
+        # it makes much more sense to use hexadecimal with 2-byte chars
+        raw = cls.bytes_bom + source.encode('utf-16-be')
+        encoding = 'hex' if bytes_encoding == 'auto' else bytes_encoding
+        return cls.from_bytes(raw, encoding)
+
+    @classmethod
+    def encode(cls, source, uni_type = type(uni_bom), isinstance=isinstance):
```

Comment:
Yeah, it's a micro-optimization.  I work hard to maintain the "fastest Python PDF processor" title :-).

You're right that you really don't get a lot out of this for a single string.  And if it were just for me, I might not do it.  But, it's a library, not a single-use program, and somebody shipping a lot of strings into a PDF using the current PdfWriter could actually notice the difference.

I was just reviewing my tokenizer yesterday and sped it up enough to make the test suite 2% faster.